File: tools/season_record.py
#!/usr/bin/env python3
"""
tools/season_record.py -- the complete 2026 season record under the
CURRENT production stack. Every graded game considered, nothing dropped.

PRICE POLICY, stated once and applied everywhere below:
    404 of 1531 graded games never had a DraftKings price captured, and
    they skew heavily to April. Two bad options and one honest one:
      * drop them        -> biases the record toward the second half
      * fill at -110     -> the silent fallback that inflated the season
                            total by ~15u in the first place
      * fill at -125     -> an EXPLICIT, LABELLED assumption
    This report fills at -125 and marks every affected bet, so you can
    always see how much of a number is assumption. -110/-140 sensitivity
    is printed at the end; if a verdict flips between them, it is an
    artefact of the fill rather than a property of the system.

WHAT IS BEING REPLAYED
    LR weights   data/lr_t1.json + lr_b1.json      (fit 2026-05-26)
    calibrator   WALK-FORWARD -- refit at each date from strictly PRIOR
                 games only, so no game is scored by a curve that has
                 seen it. This is the defensible version; the deployed
                 calibrator was fit on 2025+2026 and would be scoring
                 its own training data.
    gate         p_nrfi < _LR_STRONG_YRFI_P (read live)
    NRFI         NOT bet -- disabled in production (_LR_STRONG_NRFI_P
                 = 1.01). Reported separately for visibility.
    staking      quarter Kelly, per-bet and same-day caps, read live.

Usage:
    python tools/season_record.py
    python tools/season_record.py --games      # every bet, one per line
"""

from __future__ import annotations

import argparse
import statistics as st
import sys
from collections import defaultdict
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(ROOT))

import tracker  # noqa: E402
import mlb_first_inning_predictor as P  # noqa: E402
from tools.season_replay import load_season, payout, implied  # noqa: E402
from tools.gate_validation import walk_forward_probs, select  # noqa: E402

START = 100.0
FILL = -125.0


def build(rows, probs, fill):
    bets = select(rows, probs, side="YRFI", gate=P._LR_STRONG_YRFI_P, fill=fill)
    real = {(r["date"], r["away"], r["home"]) for r in rows if r["yrfi_odds"] is not None}
    # tag which bets used the fill
    bykey = {}
    for r in rows:
        bykey.setdefault(r["date"], []).append(r)
    for b in bets:
        b["assumed"] = abs(b["odds"] - fill) < 1e-9 if fill is not None else False
    return bets


def simulate(bets, frac=0.25, start=START):
    byday = defaultdict(list)
    for b in bets:
        byday[b["date"]].append(b)
    bank = peak = start
    mdd = 0.0
    curve = []
    staked = []
    o = tracker.KELLY_FRACTION
    tracker.KELLY_FRACTION = frac
    try:
        for d in sorted(byday):
            tracker._bankroll_cache = bank
            tracker._daily_committed = {d: 0.0}
            pnl = 0.0
            for b in byday[d]:
                s = tracker.kelly_stake_units(b["p"], str(int(b["odds"])), game_date=d) or 0.0
                b["stake"] = s
                if s <= 0:
                    continue
                staked.append(b)
                pnl += s * payout(b["odds"]) if b["win"] else -s
            bank += pnl
            peak = max(peak, bank)
            if peak > 0:
                mdd = max(mdd, (peak - bank) / peak)
            curve.append((d, bank))
    finally:
        tracker.KELLY_FRACTION = o
    return bank, 100 * mdd, curve, staked


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--games", action="store_true")
    args = ap.parse_args()

    rows, _ = load_season()
    wf = walk_forward_probs(rows)
    bets = build(rows, wf, FILL)
    bank, mdd, curve, staked = simulate(bets)

    n = len(staked)
    w = sum(1 for b in staked if b["win"])
    flat = sum(payout(b["odds"]) if b["win"] else -1.0 for b in staked)
    need = st.mean([implied(b["odds"]) for b in staked])
    assumed = sum(1 for b in staked if b["assumed"])
    days = sorted({r["date"] for r in rows})
    betdays = sorted({b["date"] for b in staked})

    print("=" * 100)
    print("  2026 SEASON RECORD -- current model + strategy, walk-forward")
    print("=" * 100)
    print(f"  gate             p_nrfi < {P._LR_STRONG_YRFI_P}")
    print(f"  staking          {tracker.KELLY_FRACTION:.2f} Kelly, "
          f"{tracker.KELLY_MAX_STAKE_FRAC:.0%}/bet, {tracker.KELLY_MAX_DAILY_FRAC:.0%}/day")
    print(f"  starting bank    {START:.0f}u")
    print(f"  missing prices   filled at {FILL:.0f}")
    print(f"  graded games     {len(rows)}   slate days {len(days)}")
    print()
    print(f"  RECORD           {w}W - {n-w}L   ({100*w/n:.1f}%)")
    print(f"  break-even need  {100*need:.1f}%   ->  edge {100*(w/n - need):+.1f} pts")
    print(f"  bets placed      {n} over {len(betdays)} days "
          f"({len(days)-len(betdays)} days with no bet)")
    print(f"  of those, priced by assumption: {assumed} ({100*assumed/n:.0f}%)")
    print()
    print(f"  FLAT 1u PROFIT   {flat:+.2f}u        (this is the EDGE)")
    print(f"  KELLY BANKROLL   {START:.0f}u -> {bank:.2f}u   =  {bank-START:+.2f}u"
          f"   ({100*(bank/START-1):+.1f}%)")
    print(f"  max drawdown     {mdd:.1f}%")
    if staked:
        ss = sorted(b["stake"] for b in staked)
        print(f"  stakes           min {ss[0]:.2f}u  median {ss[len(ss)//2]:.2f}u  max {ss[-1]:.2f}u")

    print("\n" + "-" * 100)
    print("  MONTH BY MONTH")
    print("-" * 100)
    print(f"  {'month':<9}{'bets':>6}{'W':>5}{'L':>5}{'hit%':>8}{'flat':>10}"
          f"{'bank end':>12}{'assumed':>9}")
    bym = defaultdict(list)
    for b in staked:
        bym[b["date"][:7]].append(b)
    bank_at = dict(curve)
    for m in sorted(bym):
        g = bym[m]
        mw = sum(1 for b in g if b["win"])
        mf = sum(payout(b["odds"]) if b["win"] else -1.0 for b in g)
        last = max(b["date"] for b in g)
        ma = sum(1 for b in g if b["assumed"])
        print(f"  {m:<9}{len(g):>6}{mw:>5}{len(g)-mw:>5}{100*mw/len(g):>7.1f}%"
              f"{mf:>+9.2f}u{bank_at.get(last, float('nan')):>11.2f}u{ma:>9}")

    print("\n" + "-" * 100)
    print("  PRICE-ASSUMPTION SENSITIVITY (does the verdict depend on the fill?)")
    print("-" * 100)
    print(f"  {'fill':<22}{'bets':>6}{'hit%':>8}{'flat':>10}{'Kelly bank':>13}{'maxDD':>8}")
    for f_, lbl in ((None, "real prices only"), (-110, "-110"), (-125, "-125 (used above)"),
                    (-140, "-140")):
        bb = build(rows, wf, f_)
        bk, dd, _, sk = simulate(bb)
        if not sk:
            continue
        ww = sum(1 for b in sk if b["win"])
        ff = sum(payout(b["odds"]) if b["win"] else -1.0 for b in sk)
        print(f"  {lbl:<22}{len(sk):>6}{100*ww/len(sk):>7.1f}%{ff:>+9.2f}u"
              f"{bk:>12.2f}u{dd:>7.1f}%")

    print("\n" + "-" * 100)
    print("  NRFI -- predicted but NOT bet (disabled live)")
    print("-" * 100)
    for g in (0.58, 0.60, 0.62):
        nb = select(rows, wf, side="NRFI", gate=g, fill=FILL)
        if not nb:
            continue
        nw = sum(1 for b in nb if b["win"])
        nf = sum(payout(b["odds"]) if b["win"] else -1.0 for b in nb)
        na = sum(1 for b in nb if abs(b["odds"] - FILL) < 1e-9)
        print(f"  p_nrfi >= {g:.2f}: {len(nb):>4} games  {nw}W-{len(nb)-nw}L "
              f"({100*nw/len(nb):.1f}%)  flat {nf:+.2f}u  "
              f"[{na} of {len(nb)} priced by assumption]")

    if args.games:
        print("\n" + "-" * 100)
        print("  EVERY STAKED BET")
        print("-" * 100)
        print(f"  {'date':<12}{'p(YRFI)':>9}{'odds':>7}{'stake':>8}{'result':>8}{'P&L':>9}  src")
        for b in sorted(staked, key=lambda x: x["date"]):
            pl = b["stake"] * payout(b["odds"]) if b["win"] else -b["stake"]
            print(f"  {b['date']:<12}{b['p']:>9.3f}{b['odds']:>7.0f}{b['stake']:>7.2f}u"
                  f"{'WIN' if b['win'] else 'LOSS':>8}{pl:>+8.2f}u  "
                  f"{'ASSUMED' if b['assumed'] else 'real'}")

    print("\n" + "=" * 100)
    print("  Flat 1u is the edge. The Kelly line is that same edge levered, and it")
    print("  is real only while the hit rate holds -- it unwinds at the same speed.")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: dashboard/lib/season-record.ts
import { asCumulative, type CumulativeUnits } from "./units";

/**
 * Types for data/season_record.json, written by tools/export_season_record.py.
 *
 * WHY THIS FILE EXISTS (2026-07-28)
 * ---------------------------------
 * These interfaces used to live inline in RoiPanel.tsx and described a
 * FLAT shape (side.finalBank, side.kellyProfit, day.bets, day.bankAfter).
 * The exporter now nests the simulated bankroll under `sim`, adds the
 * no-hindsight `floor`, and replaces day.bets with day.games -- a richer
 * per-game reconciliation. A stale copy of these types is not a type
 * error, it is a RUNTIME CRASH: `side.kellyProfit.toFixed(2)` on a field
 * that no longer exists throws, the panel mounts and then dies, and the
 * operator sees a blank space where his money was.
 *
 * One definition, imported everywhere. If the exporter grows a field,
 * it is added here and nowhere else.
 *
 * NOTE: `projected` and `real` are NULLABLE. make_record() returns None
 * when a side stakes nothing, which serialises as JSON null.
 */

/** One bet or one skip, as the CURRENT MODEL would have handled it. */
export interface RecDisposition {
  action:  "BET" | "SKIP";
  /** BET only. */
  stake?:   number;
  odds?:    number;
  win?:     boolean;
  pnl?:     number;
  /** BET only: price was the -125 stand-in, not a captured DK number. */
  assumed?: boolean;
  /** SKIP only. Stable machine key -- switch on this, never on `reason`. */
  code?:    "gate" | "lambda_floor" | "lambda_ceiling" | "no_price"
          | "unscored" | "kelly_no_edge" | "daily_cap" | string;
  /** SKIP only. Prose for display; wording may change, `code` will not. */
  reason?:  string;
}

/** What the LIVE system did, for games it flagged STRONG. Absent when
 *  the ledger did not flag the game at all. */
export interface RecLedger {
  strength:     string;
  placed:       boolean;
  unitsRisked:  number | null;
  odds:         number | null;
  pnl:          number | null;
}

export interface RecGame {
  game:    string;          // "TOR@WSH", or "LAD@NYY G2" for a doubleheader leg
  side:    "NRFI" | "YRFI" | string;
  modelP:  number | null;   // p(no run in the 1st)
  record:  RecDisposition;
  ledger?: RecLedger;
}

export interface RecDay {
  date:          string;
  /** The REPLAY's flat-stake P&L. NOT the ledger's -- on 2026-07-27 this
   *  is -1.00 while the ledger sums to -0.33. Never label it "you". */
  flatPnl:       number;
  /** The REPLAY's Kelly-staked P&L. Simulated money. */
  simPnl:        number;
  simBankAfter:  number | null;
  flagged:       number;    // games the live ledger called STRONG
  placed:        number;    // of those, ones actually bet
  bet:           number;    // games the current model would bet
  games:         RecGame[];
}

export interface RecFloor {
  bets: number; wins: number; losses: number;
  hitRate: number; breakEvenNeeded: number; edgePts: number;
  flatProfit: number; assumedBets: number;
  sim: { finalBank: number; profit: number; maxDrawdownPct: number;
         largestStake?: number };
}

export interface RecSide {
  label:      string;
  priceFill:  number | null;   // -125 for projected, null for real
  from:       string;
  to:         string;
  bets:       number;
  wins:       number;
  losses:     number;
  hitRate:    number;
  breakEvenNeeded: number;
  edgePts:    number;
  /** Flat 1u per bet: the raw edge with no leverage. Kept as the
   *  reference line under the Kelly headline -- the operator stakes by
   *  Kelly now, so Kelly is what the card leads with. */
  flatProfit: number;
  assumedBets: number;
  selectedBets?:    number;
  droppedZeroStake?: number;
  droppedFlatPnl?:   number;
  /** A SIMULATION: quarter-Kelly compounding an imaginary 100u bank.
   *  Never the headline; always labelled. */
  sim: {
    startBank: number; finalBank: number; profit: number;
    maxDrawdownPct: number; kellyFraction: number;
    /** What compounding actually asks of the operator. With the bank up
     *  ~10x the 10%-per-bet cap makes late stakes large in absolute
     *  units, and an average hides that. */
    largestStake?: number;
    medianStake?: number;
  };
  /** The no-hindsight lower bound. null when it stakes nothing. */
  floor:   RecFloor | null;
  monthly: { month: string; bets: number; wins: number; losses: number;
             flat: number; assumedBets: number }[];
  days:    RecDay[];
}

export interface RecFile {
  generatedUtc:   string;
  headlineMethod: string;
  floorMethod:    string;
  caveat:         string;
  gates:          { yrfi: number; nrfi: number };
  kellyFraction:  number;
  startBank:      number;
  realStart:      string;
  projected:      RecSide | null;
  real:           RecSide | null;
}

/** Guard every .toFixed(). A missing field reads as undefined, and
 *  undefined.toFixed() is the crash this file exists to prevent. */
export function isNum(v: unknown): v is number {
  return typeof v === "number" && Number.isFinite(v);
}


/** What the CURRENT model would stake on one game, at quarter-Kelly. */
export interface ReplayStake {
  action: "BET" | "SKIP";
  /** Compounded quarter-Kelly stake in units. Present only on BET. */
  stake?: number;
  /** Plain reason, present only on SKIP. */
  reason?: string;
  /** Price was the -125 stand-in rather than a captured DK number. */
  assumed?: boolean;
}

/** Key a board row the same way the exporter labels a game. */
export function replayKey(away: string, home: string, gameNumber: number | undefined,
                          side: string): string {
  const gn = gameNumber && gameNumber > 1 ? ` G${gameNumber}` : "";
  return `${away}@${home}${gn}|${side}`;
}

/**
 * Per-game quarter-Kelly stakes for one date.
 *
 * WHY (2026-07-28): the board's stake chip read `units_risked` off the
 * ledger, which is a flat 1.00 for every bet placed before Kelly went
 * live -- so browsing to April showed "staked 1.00u" on every row even
 * though the operator now sizes by Kelly. These are the SAME numbers the
 * day-reconcile table shows, so the two surfaces cannot disagree.
 *
 * Resolves against the REAL record first, then PROJECTED, because REAL
 * only starts 2026-05-07 and April exists only in the projected replay.
 */
export function replayStakesFor(
  rec: RecFile | null | undefined,
  date: string,
): Map<string, ReplayStake> {
  const out = new Map<string, ReplayStake>();
  if (!rec || !date) return out;
  const day =
    rec.real?.days.find((d) => d.date === date) ??
    rec.projected?.days.find((d) => d.date === date);
  if (!day) return out;
  for (const g of day.games) {
    const key = `${g.game}|${g.side}`;
    out.set(key, g.record.action === "BET"
      ? { action: "BET", stake: g.record.stake, assumed: g.record.assumed }
      : { action: "SKIP", reason: g.record.reason });
  }
  return out;
}


/** The system's result over an arbitrary date window.
 *
 *  EVERY P&L FIELD HERE IS BRANDED `CumulativeUnits` (2026-07-30), and
 *  that is not decoration -- `formatUnits()` refuses the brand at
 *  compile time, so `next build` fails rather than shipping a summed
 *  unit figure. See lib/units.ts for why the sum is not money. Reach
 *  for `bankStart`/`bankEnd` or the return derived from them. */
export interface ReplayWindow {
  from: string; to: string;
  bets: number; wins: number; losses: number;
  /** Quarter-Kelly P&L in units, at the bank the run had reached.
   *  Summed across days: NOT printable. `bankEnd - bankStart`. */
  pnl: CumulativeUnits;
  /** The scale-free figure. pnl as a fraction of the bank at window
   *  start -- the ONLY number that is not distorted by where the
   *  compounding run happened to be when the window opened.
   *
   *  !! DO NOT RENDER THIS FIELD. It is wrong and it has been wrong
   *  since it was written; it is kept only so that deleting it is a
   *  separate, reviewable change. `pnl` covers BOTH sides while
   *  `bankStart` is a YRFI-ONLY bank (the 2026-07-30 exporter fix
   *  stopped the simulation staking a side the system does not bet).
   *  Over the last seven days that lands on -7.7% where the bank says
   *  -5.91%, because one NRFI bet the bank never took is in the
   *  numerator. `real.days` holds 24 such bets season-wide.
   *
   *  Nothing reads it today -- RoiPanel computes its own from
   *  `yrfi.pnl / bankStart`, which is exact because the bank IS
   *  bankStart plus the YRFI P&L. Use `bankReturn(bankStart, bankEnd)`
   *  from lib/units, which cannot pick the wrong numerator. */
  pct: number | null;
  bankStart: number | null;
  bankEnd: number | null;
  /** Split by side, because only one of them is actually bet.
   *  `flat` is the same bets at one unit a bet, PER SIDE -- see the
   *  window-level `flatPnl` below for why per-side matters.
   *  Both P&L fields are summed across days, hence the brand. */
  yrfi: { bets: number; wins: number; pnl: CumulativeUnits; flat: CumulativeUnits };
  nrfi: { bets: number; wins: number; pnl: CumulativeUnits; flat: CumulativeUnits };
  /** THE EDGE, UNLEVERED -- this window at a flat 1 unit a bet, before
   *  Kelly compounding multiplies it. BOTH SIDES; use `yrfi.flat` when
   *  pairing it with a YRFI-only headline.
   *
   *  Added 2026-07-29 because the operator kept reading the levered
   *  figure as the system's performance. Showing them side by side is
   *  the difference between "the model is finding edge" and "the model
   *  is finding edge AND we are levering it hard" -- separate questions
   *  with separate risks. season_record.py's own footer: "Flat 1u is the
   *  edge. The Kelly line is that same edge levered, and it is real only
   *  while the hit rate holds."
   *
   *  2026-07-30 BUG FIX. This was summed from each day's `flatPnl`,
   *  which is a DAY-level total and therefore includes NRFI -- while
   *  every headline it sat beside is YRFI-only, because NRFI is not bet.
   *  Season-wide it printed +9.29u where the honest twin of a +144.85u
   *  YRFI headline is +12.30u; the -2.97u difference is 22 NRFI
   *  would-be bets. A figure captioned "the same bets, unlevered" that
   *  silently covered a DIFFERENT set of bets is precisely the defect
   *  this dashboard spent two days removing, reintroduced by the fix
   *  for it. Now summed per game from the same `action === "BET"` loop
   *  that produces `pnl`, so the two cannot describe different
   *  populations. */
  flatPnl: CumulativeUnits;
  /** Bets priced at the -125 stand-in because no DK price was ever
   *  captured. Load-bearing: change that assumption to -155 and the
   *  season's simulated bank falls from ~967u to ~641u. A figure built
   *  substantially on an assumed price has to say so on its face. */
  assumed: number;
}

/* ============================================================
   RE-BASED UNITS.  1 unit = 1% of bankroll, and the bankroll is
   always 100 units.  Operator decision, 2026-07-30.

   WHY ANY OF THIS EXISTS.  The replay in season_record.json compounds
   the unit COUNT: it opens a 100u bank in April and stakes a fixed
   FRACTION of whatever the bank has become, so by late July a routine
   bet is 10.00u because the bank is 217u.  Under the operator's model
   that same bet is 4.61u for everyone, because a unit is 1% of your
   own bank whatever your bank is.  The replay's raw figures describe a
   bankroll exactly one person could ever have had.

   THE CONSEQUENCE THAT BITES.  Raw per-day figures are NOT ADDABLE.
   Summing simPnl over the last seven days gives -13.17u; the bank
   actually moved 223.07 -> 209.89, which is -5.91%, i.e. -5.91u on a
   100u bank.  The naive sum is 2.2x the truth and it is the number
   every "just add the column" instinct produces.

   THE FIX, and it is one line of arithmetic: divide each day's P&L by
   the bank it OPENED with.  That ratio is bankroll-free -- it is what
   a $1k follower and a $25k follower both experienced -- and the
   ratios compound to exactly the bank ratio, so a chart drawn from
   them ends where the headline says it ends.  Verified: compounding
   the five daily returns gives -5.90u against the bank ratio's
   -5.91u, the gap being float rounding.

   WHICH SIDE.  `simBankAfter` stakes YRFI ONLY, as of the 2026-07-30
   exporter fix, because _LR_STRONG_NRFI_P = 1.01 has had NRFI
   switched off since 2026-06-07.  But `day.games` still CONTAINS
   NRFI rows -- 24 of them in the real-price window -- so anything
   that counts bets by walking games and anything that reads the bank
   are describing different populations unless the walk filters.  This
   is not hypothetical: `ReplayWindow.pct` folds NRFI into a figure it
   divides by a YRFI-only bank and lands on -7.7% where the bank says
   -5.91%.  Everything below reports YRFI and counts the off-side
   separately, so the two can never drift apart silently.
   ============================================================ */

/** One day, with its P&L expressed as a share of the bank it opened. */
export interface RebasedDay {
  date: string;
  /** P&L as a fraction of the opening bank. Bankroll-free. */
  ret: number;
  /** `ret` on a 100-unit bank -- the publishable per-night figure. */
  units: number;
  /** Bank indexed to 100 at window open, after this day settles. */
  indexed: number;
  /** The replay's own compounded figure. Kept for provenance only;
   *  never render it beside a re-based one without saying so. */
  rawPnl: number;
  bankBefore: number;
  bankAfter: number;
}

export interface RebasedWindow {
  from: string;
  to: string;
  days: RebasedDay[];
  /** The replay's compounded bank at each end of the window. */
  bankStart: number;
  bankEnd: number;
  /** THE HEADLINE: window return on a 100-unit bank. Derived from the
   *  bank ratio, never from a sum, so it cannot disagree with `days`. */
  units: number;
  /** The same figure as a fraction. */
  pct: number;
  /** YRFI only -- the side the bank actually stakes. */
  bets: number;
  wins: number;
  losses: number;
  /** Bets the record holds on a side the bank does NOT stake. Surfaced
   *  rather than dropped: a non-zero value means the record and the
   *  bank are describing different books and the reader should be told. */
  offSideBets: number;
  /** Of `bets`, how many were priced at the -125 stand-in. */
  assumed: number;
}

/**
 * The last `count` calendar days of a side's record, re-based.
 *
 * CALENDAR days, not the last N ENTRIES. `side.days` only holds dates
 * that had games, so "the last 7 entries" silently reaches back 9 or 10
 * days on a sparse stretch while the label still says a week. The
 * returned `from`/`to` are the real span so a caption can state it.
 */
export function rebaseLastDays(
  side: RecSide | null | undefined,
  count: number,
): RebasedWindow | null {
  if (!side || side.days.length === 0 || count < 1) return null;

  // A null simBankAfter means nothing settled that day; it carries no
  // bank observation, so it cannot be a point on a bank curve.
  const withBank = side.days.filter((d) => isNum(d.simBankAfter) && isNum(d.simPnl));
  if (withBank.length === 0) return null;

  const last = withBank[withBank.length - 1];
  const endMs = Date.parse(`${last.date}T00:00:00Z`);
  if (!Number.isFinite(endMs)) return null;
  const startMs = endMs - (count - 1) * 86_400_000;
  const startIso = new Date(startMs).toISOString().slice(0, 10);

  const win = withBank.filter((d) => d.date >= startIso);
  if (win.length === 0) return null;

  const days: RebasedDay[] = [];
  let indexed = 100;
  for (const d of win) {
    const bankAfter = d.simBankAfter as number;
    const rawPnl = d.simPnl;
    const bankBefore = bankAfter - rawPnl;
    // A non-positive opening bank makes the ratio meaningless (and the
    // replay is ruined anyway); treat the day as flat rather than
    // emitting an Infinity that would blow up the chart's scale.
    const ret = bankBefore > 0 ? rawPnl / bankBefore : 0;
    indexed *= 1 + ret;
    days.push({
      date: d.date, ret, units: ret * 100,
      indexed, rawPnl, bankBefore, bankAfter,
    });
  }

  const bankStart = days[0].bankBefore;
  const bankEnd = days[days.length - 1].bankAfter;
  // FROM THE RATIO, NOT FROM A SUM. This is the whole point of the
  // module: `units` and the `indexed` series are two readings of one
  // quantity, so the curve's last point IS the headline.
  const pct = bankStart > 0 ? bankEnd / bankStart - 1 : 0;

  let bets = 0, wins = 0, offSideBets = 0, assumed = 0;
  for (const d of win) {
    for (const g of d.games) {
      if (g.record.action !== "BET") continue;
      if (g.side === "NRFI") { offSideBets += 1; continue; }
      bets += 1;
      if (g.record.win === true) wins += 1;
      if (g.record.assumed) assumed += 1;
    }
  }

  return {
    from: days[0].date, to: days[days.length - 1].date,
    days, bankStart, bankEnd,
    units: pct * 100, pct,
    bets, wins, losses: bets - wins, offSideBets, assumed,
  };
}


/**
 * What the system did between two dates.
 *
 * TWO THINGS THIS MUST EXPOSE, both learned the hard way on 2026-07-28:
 *
 *  1. `pct`. The replay compounds from 100u in April, so by late July the
 *     bank is ~316u and a normal week prints as "-58.85u" -- a number
 *     that reads like a catastrophe against the operator's real 100u
 *     bankroll, where the same week is about -16u. Units are meaningless
 *     without the bank they were staked from; the percentage is not.
 *  2. The YRFI/NRFI split. NRFI betting is disabled live
 *     (_LR_STRONG_NRFI_P = 1.01), so NRFI rows in the record are TRACKED,
 *     NOT BET. Folding them into one figure reports a loss on bets that
 *     would never have been placed.
 */
export function replayWindow(
  side: RecSide | null | undefined,
  startIso: string | undefined,
  endIso: string | undefined,
): ReplayWindow | null {
  if (!side || !startIso || !endIso) return null;
  const days = side.days.filter((d) => d.date >= startIso && d.date <= endIso);
  if (days.length === 0) return null;
  let bets = 0, wins = 0, pnl = 0, assumed = 0, flatPnl = 0;
  const y = { bets: 0, wins: 0, pnl: 0, flat: 0 };
  const n = { bets: 0, wins: 0, pnl: 0, flat: 0 };
  // Accumulated as plain numbers, branded once on the way out --
  // asCumulative() at the return is the single place the "this was
  // summed across days" fact gets attached.
  for (const d of days) {
    for (const g of d.games) {
      if (g.record.action !== "BET") continue;
      const p = g.record.pnl ?? 0;
      const w = g.record.win === true;
      bets += 1; if (w) wins += 1; pnl += p;
      if (g.record.assumed) assumed += 1;
      // FLAT, PER GAME, in the same loop as pnl -- see flatPnl's note.
      // A win pays the price; a loss costs exactly one unit.
      const odds = g.record.odds;
      const payout = isNum(odds)
        ? (odds > 0 ? odds / 100 : 100 / -odds)
        : 0;
      const flat = w ? payout : -1;
      flatPnl += flat;
      const bucket = g.side === "NRFI" ? n : y;
      bucket.bets += 1; if (w) bucket.wins += 1; bucket.pnl += p;
      bucket.flat += flat;
    }
  }
  if (bets === 0) return null;
  const first = days[0], last = days[days.length - 1];
  const bankStart = isNum(first.simBankAfter) && isNum(first.simPnl)
    ? first.simBankAfter - first.simPnl : null;
  return {
    from: first.date, to: last.date,
    bets, wins, losses: bets - wins, pnl: asCumulative(pnl),
    pct: isNum(bankStart) && bankStart > 0 ? pnl / bankStart : null,
    bankStart,
    bankEnd: isNum(last.simBankAfter) ? last.simBankAfter : null,
    yrfi: { ...y, pnl: asCumulative(y.pnl), flat: asCumulative(y.flat) },
    nrfi: { ...n, pnl: asCumulative(n.pnl), flat: asCumulative(n.flat) },
    assumed, flatPnl: asCumulative(flatPnl),
  };
}
